File: Python/DesignPattern/CreationalDesignPattern/FacadePatternDemo.py
class TV:
    def on(self):
        print("TV is ON")

    def set_input(self):
        print("TV input set to HDMI")


class SoundSystem:
    def on(self):
        print("Sound System is ON")

    def set_volume(self, volume: int):
        print(f"Volume set to {volume}")


class StreamingDevice:
    def on(self):
        print("Streaming Device is ON")

    def play_movie(self, movie: str):
        print(f"Playing movie: {movie}")


class HomeTheaterFacade:
    def __init__(self):
        self._tv = TV()
        self._sound_system = SoundSystem()
        self._streaming_device = StreamingDevice()

    def watch_movie(self, movie: str):
        print("Preparing Home Theater...\n")

        self._tv.on()
        self._tv.set_input()

        self._sound_system.on()
        self._sound_system.set_volume(20)

        self._streaming_device.on()
        self._streaming_device.play_movie(movie)

        print("\nEnjoy your movie!")


if __name__ == "__main__":
    home_theater = HomeTheaterFacade()
    home_theater.watch_movie("Interstellar")

File: C++/DesignPattern/CreationalDesignPattern/FacadePatternDemo.cpp
#include <iostream>
#include <memory>
using namespace std;

// Subsystem 1
struct TV {
    void on() {
        cout << "TV is ON\n";
    }

    void setInput() {
        cout << "TV input set to HDMI\n";
    }
};

// Subsystem 2
struct SoundSystem {
    void on() {
        cout << "Sound System is ON\n";
    }

    void setVolume(int volume) {
        cout << "Volume set to " << volume << "\n";
    }
};

// Subsystem 3
struct StreamingDevice {
    void on() {
        cout << "Streaming Device is ON\n";
    }

    void playMovie(const string& movie) {
        cout << "Playing movie: " << movie << "\n";
    }
};

// Facade
class HomeTheaterFacade {
    unique_ptr<TV> tv;
    unique_ptr<SoundSystem> soundSystem;
    unique_ptr<StreamingDevice> streamingDevice;

public:
    HomeTheaterFacade()
        : tv(make_unique<TV>()),
          soundSystem(make_unique<SoundSystem>()),
          streamingDevice(make_unique<StreamingDevice>()) {}

    void watchMovie(const string& movie) {
        cout << "Preparing Home Theater...\n\n";

        tv->on();
        tv->setInput();

        soundSystem->on();
        soundSystem->setVolume(20);

        streamingDevice->on();
        streamingDevice->playMovie(movie);

        cout << "\nEnjoy your movie!\n";
    }
};

int main() {
    HomeTheaterFacade homeTheater;
    homeTheater.watchMovie("Interstellar");
    return 0;
}
